Add sensors for Tuya energy storage systems (xnyjcn)

zzysszzy opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This PR adds comprehensive support for Tuya balcony micro storage inverter devices (xnyjcn category), which are compact energy storage and solar PV inverter systems commonly used for residential solar installations. These devices combine battery storage, solar generation, and grid-tie inverter functionality.

**Key features added:**
- Complete monitoring of battery system (SOC, capacity, charge/discharge power)
- Solar PV generation monitoring (total and per-channel power, cumulative energy)
- Inverter output power and energy tracking
- Grid interaction monitoring (off-grid export energy)
- Device configuration controls (backup reserve, power limits, work modes)
- System information display (serial number, battery capacity)
- Proper energy unit handling with suggested measurements

This enables users to fully monitor and control their Tuya-based home energy storage systems, providing insights into solar generation, battery usage, and energy flow patterns.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: adds support for Tuya energy storage devices
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

**Technical details:**
- Added support for device category `xnyjcn` (Balcony Micro Storage Inverter)
- Extended `homeassistant/components/tuya/sensor.py` with energy and power monitoring
- Enhanced `homeassistant/components/tuya/number.py` with configuration controls
- Updated `homeassistant/components/tuya/select.py` with work mode selection
- Added switch control in `homeassistant/components/tuya/switch.py` for power limiting
- Comprehensive DPCode definitions for energy storage systems
- Complete translation support for all energy storage entities

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

zzysszzy committed:
Add Balcony Micro Storage Inverter(xnyjcn) entity

zzysszzy committed:
Add Balcony Micro Storage Inverter(xnyjcn) entity

zzysszzy committed:
Optimize comments according to PR feedback

zzysszzy committed:
Optimize comments according to PR feedback

epenet reviewed (changes requested):
Snapshot tests are still missing (new fixture missing in DEVICE_MOCKS)
DPCode are still not ordered correctly (alphabetical order)
pv_power_channel_1/2 should use indexed translation placeholder

zzysszzy committed:
Sort DPCODE alphabetically

zzysszzy committed:
Sort DPCODE alphabetically

zzysszzy commented:
> Snapshot tests are still missing (new fixture missing in DEVICE_MOCKS) DPCode are still not ordered correctly (alphabetical order) pv_power_channel_1/2 should use indexed translation placeholder

Do i need to run "pytest /tests/components/tuya --snapshot-update" to update all ambr snapshot files? Or just fix DEVICE_MOCKS

zzysszzy committed:
Optimize according to PR feedback

zzysszzy committed:
Optimize according to PR feedback

epenet reviewed (changes requested):
Tests are failing. Maybe you need to merge dev (or rebase) before updating snapshots

zzysszzy committed:
update snapshot

zzysszzy committed:
update snapshot

epenet commented on homeassistant/components/tuya/const.py:
```diff
@@ -395,10 +411,10 @@ class DPCode(StrEnum):
     WARM = "warm"  # Heat preservation
     WARM_TIME = "warm_time"  # Heat preservation time
     WATER = "water"
+    WATER_LEVEL = "water_level"
```
Just to make it clearer, can you create a small preliminary PR to sort the existing keys?

epenet replied:
After your first PR is merged the CI runs will be enabled automatically... so it's worth creating your first PR really small

Currently thus PR is too big for me to review on mobile phone so unless it is picked up by another reviewer in the mean time...

zzysszzy replied:
> After your first PR is merged the CI runs will be enabled automatically... so it's worth creating your first PR really small
> 
> Currently thus PR is too big for me to review on mobile phone so unless it is picked up by another reviewer in the mean time...

I've completed the modifications - only the newly added DPCODEs have been alphabetically sorted, with all existing DPCODEs remaining unchanged.
Please check it again.

zzysszzy committed:
Merge branch 'feature/tuya-energy-storage' of github.com:zzysszzy/HomeAssiatant-core into feature/tuya-energy-storage

zzysszzy committed:
Sort only newly added DPCode alphabetically

Only sort the newly added DPCode entries while keeping the original
DPCode order unchanged to minimize the diff.

zzysszzy committed:
Update homeassistant/components/tuya/number.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

zzysszzy committed:
Update homeassistant/components/tuya/number.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

zzysszzy committed:
merge dev

zzysszzy committed:
change work_mode to inverter_work_mode

zzysszzy committed:
Merge branch 'feature/tuya-energy-storage' of github.com:zzysszzy/HomeAssiatant-core into feature/tuya-energy-storage

zzysszzy committed:
merge dev

zzysszzy committed:
fix ruff

zzysszzy committed:
update snapshot

zzysszzy committed:
merge dev

zzysszzy commented on homeassistant/components/tuya/select.py:
```diff
@@ -352,6 +352,15 @@
+    "xnyjcn": (
+        SelectEntityDescription(
+            key=DPCode.WORK_MODE,
+            translation_key="work_mode",
```
It's ok now, please check it again.

zzysszzy committed:
Merge branch 'dev' into feature/tuya-energy-storage

zzysszzy committed:
move serial_number from sensor entity to device info

epenet commented on homeassistant/components/tuya/sensor.py:
```diff
@@ -1279,7 +1280,96 @@ class TuyaSensorEntityDescription(SensorEntityDescription):
+            key=DPCode.SERIAL_NUMBER,
+            translation_key="serial_number",
+            entity_category=EntityCategory.DIAGNOSTIC,
+        ),
```
> I've moved the serial number to the device info section. Please verify if this is correct.

I suggest to remove it completely, and create a follow-up PR later if you really feel it is needed.
Same for other functionnality that is not straightforward and may need "debating"

zzysszzy committed:
update snapshot

zzysszzy committed:
remove serial number

zzysszzy committed:
merge dev

zzysszzy committed:
remove serial number from consts

zzysszzy committed:
fix test_init.ambr for xnyjcn

zzysszzy committed:
Trigger CI

frenck commented on homeassistant/components/tuya/sensor.py:
```diff
@@ -1279,7 +1280,96 @@ class TuyaSensorEntityDescription(SensorEntityDescription):
+            translation_key="serial_number",
+            entity_category=EntityCategory.DIAGNOSTIC,
+        ),
+        TuyaSensorEntityDescription(
+            key=DPCode.BATTERY_CAPACITY,
+            translation_key="battery_capacity",
+            device_class=SensorDeviceClass.ENERGY,
+            entity_category=EntityCategory.DIAGNOSTIC,
+        ),
+    ),
```
Sure, but they can do so statically. You are putting this in a state machine, which is designed to handle state changes and act on that; not to log configuration data in.

zzysszzy replied:
OK. I'll remove it now and only submit a new PR to add it when we receive confirmed user demand.

zzysszzy committed:
remove battery capacity

zzysszzy commented:
Are there any other issues with this PR? @frenck @epenet

zzysszzy committed:
merge dev

epenet commented on homeassistant/components/tuya/sensor.py:
```diff
@@ -1332,7 +1333,85 @@ class TuyaSensorEntityDescription(SensorEntityDescription):
+    "xnyjcn": (
+        TuyaSensorEntityDescription(
+            key=DPCode.CURRENT_SOC,
+            translation_key="battery_soc",
```
Is the translation key necessary for the battery soc?

zzysszzy replied:
Do you mean the translation_key doesn't need to be specified?

epenet commented on homeassistant/components/tuya/sensor.py:
```diff
@@ -1332,7 +1333,85 @@ class TuyaSensorEntityDescription(SensorEntityDescription):
+            translation_key="cumulative_energy_output_inv",
+            device_class=SensorDeviceClass.ENERGY,
+            state_class=SensorStateClass.TOTAL_INCREASING,
+            suggested_unit_of_measurement=UnitOfEnergy.KILO_WATT_HOUR,
```
I don't think suggested_unit_of_measurement is needed on any of the sensors
It should be detected from the Tuya SDK

zzysszzy committed:
remove suggest unit

zzysszzy committed:
merge dev

epenet commented on homeassistant/components/tuya/number.py:
```diff
@@ -399,6 +399,21 @@
+    "xnyjcn": (
+        NumberEntityDescription(
+            key=DPCode.BACKUP_RESERVE,
+            translation_key="backup_reserve",
```
This is the backup reserve of what? The battery storage?

zzysszzy replied:
This represents the battery backup power percentage, the minimum SOC (State of Charge) value for battery discharge.

epenet commented on homeassistant/components/tuya/number.py:
```diff
@@ -399,6 +399,21 @@
+        ),
+        NumberEntityDescription(
+            key=DPCode.OUTPUT_POWER_LIMIT,
+            translation_key="output_power_limit",
```
This is the output power limit of what? The PV to grid? the battery storage to home? something else?

zzysszzy replied:
This represents the inverter output power limit, the maximum power limit for converting DC from PV and batteries to AC output.

epenet reviewed (commented):
I think the reason this PR is still not merged is that it's still quite big and some DP codes / translation codes are still confusing.

You could start with a single platform (eg. `number` or `switch`) and add the others after.

epenet commented:
You can also create a PR just for the fixtures - no code changes (see #151185)

zzysszzy committed:
merge dev

zzysszzy committed:
remove select switch number entity

zzysszzy commented:
Hi @epenet  
This is another PR for the xnyjcn category. Could you please review it as well? Thank you.

zzysszzy committed:
Merge branch 'dev' into feature/tuya-energy-storage

epenet commented on tests/components/tuya/snapshots/test_init.ambr:
```diff
@@ -2758,6 +2758,37 @@
     'via_device_id': None,
   })
 # ---
+# name: test_device_registry[gbq8kiahk57ct0bpncjynx]
+  DeviceRegistryEntrySnapshot({
```
I think there is a bad merge: gbq8kiahk57ct0bpncjynx is now duplicated in this snapshot file

zzysszzy replied:
Yes, it was indeed a duplicate. I’ve resolved it

epenet commented on homeassistant/components/tuya/sensor.py:
```diff
@@ -1365,7 +1365,80 @@ class TuyaSensorEntityDescription(SensorEntityDescription):
+        ),
+        TuyaSensorEntityDescription(
+            key=DPCode.CUML_E_EXPORT_OFFGRID1,
+            translation_key="cuml_e_export_offgrid1",
```
Can you please review the translation keys?
As mentionned in other PRs, it would be great to have the keys match more closely with the english text (not with the DPCode)

zzysszzy replied:
> Can you please review all the translation keys? As mentionned in other PRs, it would be great to have the keys match more closely with the english text (not with the DPCode)

It's ok now, please check it again.

zzysszzy committed:
Clean up duplicate gbq8kiahk57ct0bpncjynx

zzysszzy committed:
Optimize translation_key

zzysszzy committed:
Optimize translation_key

zzysszzy committed:
Optimize translation_key

zzysszzy committed:
Optimize translation_key

zzysszzy commented on homeassistant/components/tuya/strings.json:
```diff
@@ -582,6 +582,36 @@
+      "battery_soc": {
+        "name": "Battery SOC"
+      },
+      "battery_power": {
+        "name": "Battery power (+ discharge, - charge)"
```
I’ll remove the positive/negative descriptions since different manufacturers may use different conventions. Most battery devices represent discharge with positive values.

epenet replied:
If we are sure of the direction, you could go with "Battery power output"
If we are unsure of the direction, then we should stay generic "Battery power"

zzysszzy replied:
It's "Battery power" now.

zzysszzy committed:
Optimize translation_key

zzysszzy committed:
Optimize translation_key

zzysszzy committed:
Optimize translation_key

zzysszzy commented on homeassistant/components/tuya/strings.json:
```diff
@@ -582,6 +582,36 @@
+        "name": "Total battery charge energy"
+      },
+      "total_offgrid_port1_output_energy": {
+        "name": "Total off-grid port 1 output energy"
```
Now it is "Lifetime off-grid output energy". Is that ok?

zzysszzy replied:
Or it should be "Lifetime off-grid port energy"?

NoRi2909 replied:
I assume that the energy coming out of this port is part of the "inverter output energy".

Then I'd use "port" here, this inherently explains that it is coming from that connection.

zzysszzy replied:
> Is there (/will there be) a port2/port3/port4 ?

Currently, there is only one off-grid port. So i removed the number 1

zzysszzy replied:
> I assume that the energy coming out of this port is part of the "inverter output energy".
> 
> Then I'd use "port" here, this inherently explains that it is coming from that connection.

Yes, it's part of the "inverter output energy". I'll use "Lifetime off-grid port energy".

zzysszzy committed:
Optimize translation_key

NoRi2909 reviewed (approved):
Strings look OK to me now.

epenet reviewed (approved):
LGTM 👍

zzysszzy commented:
Hi @frenck , could you please review this PR again to see if there are any remaining issues?

zzysszzy committed:
Merge branch 'dev' into feature/tuya-energy-storage

zzysszzy committed:
merge dev

zzysszzy commented:
> A conflict has appeared

It's fixed now.

zzysszzy commented:
@epenet @NoRi2909 Since you've already approved this PR, could you help push it forward or suggest next steps?
